File: LastLogged/Services/SupabaseService.swift
import Foundation
import SwiftData
import Supabase
import Network
import UIKit

@Observable
final class SupabaseService {
    static let shared = SupabaseService()

    // MARK: - Configuration (read from AppSecrets)

    private static let supabaseURL = URL(string: AppSecrets.supabaseURL)!
    private static let supabaseAnonKey = AppSecrets.supabaseAnonKey

    let client: SupabaseClient

    // MARK: - State

    private(set) var isSyncing = false
    private(set) var lastSyncedAt: Date?
    private(set) var isSignedIn = false
    private(set) var currentUserEmail: String?
    private(set) var isNetworkAvailable = true
    private var cachedUserId: UUID?
    private var syncDebounceTask: Task<Void, Never>?
    private let syncDebounceInterval: TimeInterval = 2.0
    private var hasPendingSync = false
    private var pendingSyncModelContext: ModelContext?

    // Network monitoring
    private let networkMonitor = NWPathMonitor()
    private let networkQueue = DispatchQueue(label: "com.pearsonmedia.lastlogged.network")

    // Retry config
    private static let maxRetryAttempts = 4
    private static let baseRetryDelay: TimeInterval = 2.0

    // UserDefaults keys
    private static let lastSyncTimestampKey = "com.pearsonmedia.lastlogged.lastSyncTimestamp"
    private static let hasPendingSyncKey = "com.pearsonmedia.lastlogged.hasPendingSync"

    private var lastSyncTimestamp: Date? {
        get { UserDefaults.standard.object(forKey: Self.lastSyncTimestampKey) as? Date }
        set { UserDefaults.standard.set(newValue, forKey: Self.lastSyncTimestampKey) }
    }

    // MARK: - Initialization

    private init() {
        client = SupabaseClient(
            supabaseURL: Self.supabaseURL,
            supabaseKey: Self.supabaseAnonKey
        )
        hasPendingSync = UserDefaults.standard.bool(forKey: Self.hasPendingSyncKey)
        startNetworkMonitoring()
    }

    // MARK: - Network Monitoring

    private func startNetworkMonitoring() {
        networkMonitor.pathUpdateHandler = { [weak self] path in
            let wasAvailable = self?.isNetworkAvailable ?? false
            let isNowAvailable = path.status == .satisfied

            Task { @MainActor in
                self?.isNetworkAvailable = isNowAvailable

                // Auto-sync when network returns and there are pending changes
                if !wasAvailable && isNowAvailable && self?.hasPendingSync == true {
                    if let context = self?.pendingSyncModelContext {
                        self?.syncOnForeground(modelContext: context)
                    }
                }
            }
        }
        networkMonitor.start(queue: networkQueue)
    }

    // MARK: - Retry Helper

    private func withRetry<T>(
        context: String,
        operation: () async throws -> T
    ) async throws -> T {
        var lastError: Error?
        for attempt in 0..<Self.maxRetryAttempts {
            do {
                return try await operation()
            } catch {
                lastError = error
                if attempt < Self.maxRetryAttempts - 1 {
                    let delay = Self.baseRetryDelay * pow(2.0, Double(attempt))
                    try? await Task.sleep(for: .seconds(delay))
                }
            }
        }
        AnalyticsService.shared.trackError("sync_\(context)_exhausted", error: lastError!)
        throw lastError!
    }

    // MARK: - Session Restore

    func restoreSession() async {
        do {
            let session = try await client.auth.session
            isSignedIn = true
            currentUserEmail = session.user.email
        } catch {
            isSignedIn = false
            currentUserEmail = nil
            cachedUserId = nil
        }
    }

    // MARK: - Apple Sign In

    func signInWithApple(idToken: String, nonce: String) async throws {
        cachedUserId = nil
        let session = try await client.auth.signInWithIdToken(
            credentials: .init(
                provider: .apple,
                idToken: idToken,
                nonce: nonce
            )
        )
        isSignedIn = true
        currentUserEmail = session.user.email
        await createOrUpdateUserProfile(session: session)
    }

    // MARK: - Email/Password Auth

    func signUpEmail(email: String, password: String) async throws {
        cachedUserId = nil
        let session = try await client.auth.signUp(
            email: email,
            password: password
        ).session
        if let session {
            isSignedIn = true
            currentUserEmail = session.user.email
            await createOrUpdateUserProfile(session: session)
        }
    }

    func signInEmail(email: String, password: String) async throws {
        cachedUserId = nil
        let session = try await client.auth.signIn(
            email: email,
            password: password
        )
        isSignedIn = true
        currentUserEmail = session.user.email
        await createOrUpdateUserProfile(session: session)
    }

    // MARK: - Password Reset

    func resetPassword(email: String) async throws {
        try await client.auth.resetPasswordForEmail(email)
    }

    // MARK: - Account Deletion

    func deleteAccount() async throws {
        let session = try await client.auth.session
        let token = session.accessToken

        // Call server-side delete endpoint
        let edgeFunctionURL = Self.supabaseURL.appendingPathComponent("delete-account")
        var request = URLRequest(url: edgeFunctionURL)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "DeleteAccount", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        if httpResponse.statusCode != 200 {
            let body = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw NSError(domain: "DeleteAccount", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "Account deletion failed: \(body)"])
        }

        // Sign out locally
        try? await client.auth.signOut()
        isSignedIn = false
        currentUserEmail = nil
        cachedUserId = nil

        AnalyticsService.shared.trackEvent("account_deleted")
    }

    // MARK: - Sign Out

    func signOut() async throws {
        try await client.auth.signOut()
        isSignedIn = false
        currentUserEmail = nil
        cachedUserId = nil
    }

    // MARK: - User Profile

    private func createOrUpdateUserProfile(session: Session) async {
        let authId = session.user.id
        let displayName = session.user.userMetadata["full_name"]?.stringValue
            ?? session.user.email
            ?? "User"
        let tier = RevenueCatService.shared.subscriptionTier.rawValue

        let row = UserProfileRow(
            authId: authId,
            displayName: displayName,
            subscriptionTier: tier
        )
        do {
            try await client.from("users")
                .upsert(row, onConflict: "auth_id")
                .execute()
        } catch {
            AnalyticsService.shared.trackError("profile_upsert_failed", error: error)
        }
    }

    // MARK: - Device Token Registration

    func registerDeviceToken(_ token: String) async {
        guard let userId = await currentUserId else { return }

        let deviceName = await UIDevice.current.name
        let row = DeviceTokenRow(
            userId: userId,
            deviceToken: token,
            deviceName: deviceName,
            platform: "ios",
            lastSeenAt: Date()
        )
        do {
            try await client.from("user_devices")
                .upsert(row, onConflict: "device_token")
                .execute()
        } catch {
            AnalyticsService.shared.trackError("device_token_register_failed", error: error)
        }
    }

    // MARK: - Auth State

    var isAuthenticated: Bool {
        get async {
            do {
                _ = try await client.auth.session
                return true
            } catch {
                return false
            }
        }
    }

    var currentUserId: UUID? {
        get async {
            // Return cached value if available
            if let cachedUserId {
                return cachedUserId
            }
            do {
                let session = try await client.auth.session
                let authId = session.user.id
                // Look up internal user ID from users table
                let rows: [UserRow] = try await client.from("users")
                    .select("id")
                    .eq("auth_id", value: authId.uuidString)
                    .execute()
                    .value
                let userId = rows.first?.id
                cachedUserId = userId
                return userId
            } catch {
                return nil
            }
        }
    }

    // MARK: - Sync Trigger (Debounced)

    func scheduleSyncAfterWrite(modelContext: ModelContext) {
        // Mark pending sync (persists across app restarts)
        hasPendingSync = true
        pendingSyncModelContext = modelContext
        UserDefaults.standard.set(true, forKey: Self.hasPendingSyncKey)

        // Debounce: only delays the trigger, does not cancel pending data
        syncDebounceTask?.cancel()
        let context = modelContext
        syncDebounceTask = Task { @MainActor in
            try? await Task.sleep(for: .seconds(syncDebounceInterval))
            guard !Task.isCancelled else { return }
            await sync(modelContext: context)
        }
    }

    func syncOnForeground(modelContext: ModelContext) {
        pendingSyncModelContext = modelContext
        Task { @MainActor in
            await sync(modelContext: modelContext)
        }
    }

    // MARK: - Full Sync (Push then Pull)

    @MainActor
    func sync(modelContext: ModelContext) async {
        guard await isAuthenticated else { return }
        guard !isSyncing else { return }
        guard isNetworkAvailable else {
            hasPendingSync = true
            UserDefaults.standard.set(true, forKey: Self.hasPendingSyncKey)
            return
        }

        isSyncing = true
        defer { isSyncing = false }

        guard let userId = await currentUserId else {
            AnalyticsService.shared.trackError("sync_no_user_id", error: NSError(domain: "Sync", code: -1, userInfo: [NSLocalizedDescriptionKey: "Could not resolve user ID"]))
            return
        }

        await pushCategories(modelContext: modelContext, userId: userId)
        await pushTrackerItems(modelContext: modelContext, userId: userId)
        await pushCompletionLogs(modelContext: modelContext, userId: userId)

        await pullCategories(modelContext: modelContext, userId: userId)
        await pullTrackerItems(modelContext: modelContext, userId: userId)
        await pullCompletionLogs(modelContext: modelContext, userId: userId)

        lastSyncedAt = Date()
        lastSyncTimestamp = Date()

        // Clear pending flag on successful sync
        hasPendingSync = false
        UserDefaults.standard.set(false, forKey: Self.hasPendingSyncKey)

        do {
            try modelContext.save()
        } catch {
            AnalyticsService.shared.trackError("sync_save_failed", error: error)
        }
    }

    // MARK: - Push Categories

    private func pushCategories(modelContext: ModelContext, userId: UUID) async {
        let descriptor = FetchDescriptor<TrackerCategory>()
        guard let categories = try? modelContext.fetch(descriptor) else { return }

        for category in categories {
            let row = TrackerCategoryRow(
                id: category.id,
                userId: userId,
                name: category.name,
                iconName: category.iconName,
                colorHex: category.colorHex,
                sortOrder: category.sortOrder,
                isDefault: category.isDefault
            )
            do {
                try await withRetry(context: "push_category") {
                    try await client.from("tracker_categories")
                        .upsert(row, onConflict: "id")
                        .execute()
                }
            } catch {
                AnalyticsService.shared.trackError("push_category_failed", error: error)
            }
        }
    }

    // MARK: - Push Tracker Items

    private func pushTrackerItems(modelContext: ModelContext, userId: UUID) async {
        let descriptor = FetchDescriptor<TrackerItem>(
            predicate: #Predicate { $0.syncStatus != .synced }
        )
        guard let items = try? modelContext.fetch(descriptor) else { return }

        for item in items {
            let row = TrackerItemRow(
                id: item.id,
                userId: userId,
                name: item.name,
                categoryId: item.categoryId,
                reminderIntervalDays: item.reminderIntervalDays,
                lastCompletedAt: item.lastCompletedAt,
                createdAt: item.createdAt,
                sortOrder: item.sortOrder,
                iconName: item.iconName,
                isArchived: item.isArchived,
                syncStatus: "synced",
                updatedAt: item.updatedAt
            )
            do {
                try await withRetry(context: "push_item") {
                    try await client.from("tracker_items")
                        .upsert(row, onConflict: "id")
                        .execute()
                }
                // Only mark synced AFTER confirmed server push
                item.syncStatus = .synced
            } catch {
                // Keep as .pending — will retry on next sync
                AnalyticsService.shared.trackError("push_item_failed", error: error)
            }
        }
    }

    // MARK: - Push Completion Logs

    private func pushCompletionLogs(modelContext: ModelContext, userId: UUID) async {
        let descriptor = FetchDescriptor<CompletionLog>()
        guard let logs = try? modelContext.fetch(descriptor) else { return }

        for log in logs {
            let row = CompletionLogRow(
                id: log.id,
                userId: userId,
                trackerItemId: log.trackerItemId,
                completedAt: log.completedAt,
                notes: log.notes
            )
            do {
                try await withRetry(context: "push_log") {
                    try await client.from("completion_logs")
                        .upsert(row, onConflict: "id")
                        .execute()
                }
            } catch {
                AnalyticsService.shared.trackError("push_log_failed", error: error)
            }
        }
    }

    // MARK: - Pull Categories

    private func pullCategories(modelContext: ModelContext, userId: UUID) async {
        do {
            var query = client.from("tracker_categories")
                .select()
                .eq("user_id", value: userId.uuidString)

            if let since = lastSyncTimestamp {
                let formatter = ISO8601DateFormatter()
                formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
                query = query.gte("updated_at", value: formatter.string(from: since))
            }

            let rows: [TrackerCategoryRow] = try await query.execute().value

            for row in rows {
                let descriptor = FetchDescriptor<TrackerCategory>(
                    predicate: #Predicate { $0.id == row.id }
                )
                let existing = try? modelContext.fetch(descriptor).first

                if let existing {
                    // Last-write-wins: remote is newer (we just pulled)
                    existing.name = row.name
                    existing.iconName = row.iconName
                    existing.colorHex = row.colorHex
                    existing.sortOrder = row.sortOrder
                    existing.isDefault = row.isDefault
                } else {
                    let category = TrackerCategory(
                        id: row.id,
                        name: row.name,
                        iconName: row.iconName,
                        colorHex: row.colorHex,
                        sortOrder: row.sortOrder,
                        isDefault: row.isDefault
                    )
                    modelContext.insert(category)
                }
            }
        } catch {
            AnalyticsService.shared.trackError("pull_categories_failed", error: error)
        }
    }

    // MARK: - Pull Tracker Items

    private func pullTrackerItems(modelContext: ModelContext, userId: UUID) async {
        do {
            var query = client.from("tracker_items")
                .select()
                .eq("user_id", value: userId.uuidString)

            if let since = lastSyncTimestamp {
                let formatter = ISO8601DateFormatter()
                formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
                query = query.gte("updated_at", value: formatter.string(from: since))
            }

            let rows: [TrackerItemRow] = try await query.execute().value

            for row in rows {
                let descriptor = FetchDescriptor<TrackerItem>(
                    predicate: #Predicate { $0.id == row.id }
                )
                let existing = try? modelContext.fetch(descriptor).first

                if let existing {
                    let remoteUpdatedAt = row.updatedAt ?? .distantPast
                    let localUpdatedAt = existing.updatedAt

                    if localUpdatedAt > remoteUpdatedAt && existing.syncStatus != .synced {
                        // Local is newer — keep local version, mark as pending for next push
                        existing.syncStatus = .pending
                        AnalyticsService.shared.trackSyncConflict(itemId: existing.id)
                    } else {
                        // Remote is newer or equal — overwrite local
                        existing.name = row.name
                        existing.categoryId = row.categoryId
                        existing.reminderIntervalDays = row.reminderIntervalDays
                        existing.lastCompletedAt = row.lastCompletedAt
                        existing.sortOrder = row.sortOrder
                        existing.iconName = row.iconName
                        existing.isArchived = row.isArchived
                        existing.syncStatus = .synced
                        existing.updatedAt = remoteUpdatedAt
                    }
                } else {
                    let item = TrackerItem(
                        id: row.id,
                        name: row.name,
                        categoryId: row.categoryId,
                        reminderIntervalDays: row.reminderIntervalDays,
                        lastCompletedAt: row.lastCompletedAt,
                        createdAt: row.createdAt,
                        sortOrder: row.sortOrder,
                        iconName: row.iconName,
                        isArchived: row.isArchived,
                        syncStatus: .synced,
                        updatedAt: row.updatedAt ?? Date()
                    )
                    modelContext.insert(item)
                }
            }
        } catch {
            AnalyticsService.shared.trackError("pull_items_failed", error: error)
        }
    }

    // MARK: - Pull Completion Logs

    private func pullCompletionLogs(modelContext: ModelContext, userId: UUID) async {
        do {
            var query = client.from("completion_logs")
                .select()
                .eq("user_id", value: userId.uuidString)

            if let since = lastSyncTimestamp {
                let formatter = ISO8601DateFormatter()
                formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
                query = query.gte("created_at", value: formatter.string(from: since))
            }

            let rows: [CompletionLogRow] = try await query.execute().value

            for row in rows {
                let descriptor = FetchDescriptor<CompletionLog>(
                    predicate: #Predicate { $0.id == row.id }
                )
                let existing = try? modelContext.fetch(descriptor).first

                if existing == nil {
                    let log = CompletionLog(
                        id: row.id,
                        trackerItemId: row.trackerItemId,
                        completedAt: row.completedAt,
                        notes: row.notes
                    )
                    modelContext.insert(log)
                }
            }
        } catch {
            AnalyticsService.shared.trackError("pull_logs_failed", error: error)
        }
    }
}

// MARK: - Supabase Row DTOs

private struct UserRow: Decodable {
    let id: UUID
}

struct UserProfileRow: Encodable {
    let authId: UUID
    let displayName: String
    let subscriptionTier: String

    enum CodingKeys: String, CodingKey {
        case authId = "auth_id"
        case displayName = "display_name"
        case subscriptionTier = "subscription_tier"
    }
}

struct TrackerCategoryRow: Codable {
    let id: UUID
    let userId: UUID
    let name: String
    let iconName: String
    let colorHex: String
    let sortOrder: Int
    let isDefault: Bool

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case name
        case iconName = "icon_name"
        case colorHex = "color_hex"
        case sortOrder = "sort_order"
        case isDefault = "is_default"
    }
}

struct TrackerItemRow: Codable {
    let id: UUID
    let userId: UUID
    let name: String
    let categoryId: UUID
    let reminderIntervalDays: Int?
    let lastCompletedAt: Date?
    let createdAt: Date
    let sortOrder: Int
    let iconName: String
    let isArchived: Bool
    let syncStatus: String
    let updatedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case name
        case categoryId = "category_id"
        case reminderIntervalDays = "reminder_interval_days"
        case lastCompletedAt = "last_completed_at"
        case createdAt = "created_at"
        case sortOrder = "sort_order"
        case iconName = "icon_name"
        case isArchived = "is_archived"
        case syncStatus = "sync_status"
        case updatedAt = "updated_at"
    }
}

struct CompletionLogRow: Codable {
    let id: UUID
    let userId: UUID
    let trackerItemId: UUID
    let completedAt: Date
    let notes: String?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case trackerItemId = "tracker_item_id"
        case completedAt = "completed_at"
        case notes
    }
}

struct DeviceTokenRow: Encodable {
    let userId: UUID
    let deviceToken: String
    let deviceName: String
    let platform: String
    let lastSeenAt: Date

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case deviceToken = "device_token"
        case deviceName = "device_name"
        case platform
        case lastSeenAt = "last_seen_at"
    }
}

File: android/app/src/main/java/com/pearsonmedia/lastlogged/service/SupabaseService.kt
package com.pearsonmedia.lastlogged.service

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.util.Log
import com.pearsonmedia.lastlogged.BuildConfig
import com.pearsonmedia.lastlogged.data.local.dao.CompletionLogDao
import com.pearsonmedia.lastlogged.data.local.dao.TrackerCategoryDao
import com.pearsonmedia.lastlogged.data.local.dao.TrackerItemDao
import com.pearsonmedia.lastlogged.data.local.entity.SyncStatus
import com.pearsonmedia.lastlogged.data.local.entity.TrackerItem
import dagger.hilt.android.qualifiers.ApplicationContext
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.gotrue.Auth
import io.github.jan.supabase.gotrue.auth
import io.github.jan.supabase.gotrue.providers.Google
import io.github.jan.supabase.gotrue.providers.builtin.Email
import io.github.jan.supabase.gotrue.providers.builtin.IDToken
import io.github.jan.supabase.postgrest.Postgrest
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
data class TrackerItemRow(
    val id: String,
    val name: String,
    @SerialName("category_id") val categoryId: String? = null,
    @SerialName("reminder_interval_days") val reminderIntervalDays: Int = 30,
    @SerialName("last_completed_at") val lastCompletedAt: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("sort_order") val sortOrder: Int = 0,
    @SerialName("icon_name") val iconName: String = "checklist",
    @SerialName("is_archived") val isArchived: Boolean = false,
    @SerialName("updated_at") val updatedAt: String? = null
)

@Serializable
data class TrackerCategoryRow(
    val id: String,
    val name: String,
    @SerialName("icon_name") val iconName: String = "category",
    @SerialName("color_hex") val colorHex: String = "#4f46e5",
    @SerialName("sort_order") val sortOrder: Int = 0,
    @SerialName("is_default") val isDefault: Boolean = false
)

@Serializable
data class CompletionLogRow(
    val id: String,
    @SerialName("tracker_item_id") val trackerItemId: String,
    @SerialName("completed_at") val completedAt: String? = null,
    val notes: String? = null
)

@Singleton
class SupabaseService @Inject constructor(
    @ApplicationContext private val context: Context,
    private val trackerItemDao: TrackerItemDao,
    private val completionLogDao: CompletionLogDao,
    private val trackerCategoryDao: TrackerCategoryDao,
    private val secureStorageService: SecureStorageService
) {
    companion object {
        private const val TAG = "SupabaseService"
        private const val SYNC_DEBOUNCE_MS = 2000L
        private const val MAX_RETRY_ATTEMPTS = 4
        private val RETRY_DELAYS_MS = longArrayOf(2000, 4000, 8000, 16000)
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val syncMutex = Mutex()
    private var debouncedSyncJob: Job? = null

    val client: SupabaseClient = createSupabaseClient(
        supabaseUrl = BuildConfig.SUPABASE_URL,
        supabaseKey = BuildConfig.SUPABASE_ANON_KEY
    ) {
        install(Auth)
        install(Postgrest)
    }

    private val _isSignedIn = MutableStateFlow(false)
    val isSignedIn: StateFlow<Boolean> = _isSignedIn.asStateFlow()

    private val _currentUserEmail = MutableStateFlow<String?>(null)
    val currentUserEmail: StateFlow<String?> = _currentUserEmail.asStateFlow()

    private val _isConnected = MutableStateFlow(true)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _isSyncing = MutableStateFlow(false)
    val isSyncing: StateFlow<Boolean> = _isSyncing.asStateFlow()

    private val _syncError = MutableStateFlow<String?>(null)
    val syncError: StateFlow<String?> = _syncError.asStateFlow()

    private val _sessionExpired = MutableStateFlow(false)
    val sessionExpired: StateFlow<Boolean> = _sessionExpired.asStateFlow()

    private var cachedUserId: String? = null

    init {
        setupNetworkMonitoring()
        setupAuthStateListener()
    }

    // --- Network Monitoring ---

    private fun setupNetworkMonitoring() {
        val connectivityManager = context.getSystemService(ConnectivityManager::class.java)
        val request = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()

        connectivityManager.registerNetworkCallback(request, object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                _isConnected.value = true
                // Trigger sync when connectivity returns
                scheduleDebouncedSync()
            }

            override fun onLost(network: Network) {
                _isConnected.value = false
            }
        })
    }

    // --- Auth State Listener ---

    private fun setupAuthStateListener() {
        scope.launch {
            client.auth.sessionStatus.collect { status ->
                when (status) {
                    is io.github.jan.supabase.gotrue.SessionStatus.Authenticated -> {
                        _isSignedIn.value = true
                        _sessionExpired.value = false
                        _currentUserEmail.value = status.session.user?.email
                        cachedUserId = null // Clear cache, will be refreshed on next sync
                    }
                    is io.github.jan.supabase.gotrue.SessionStatus.NotAuthenticated -> {
                        _isSignedIn.value = false
                        _currentUserEmail.value = null
                        cachedUserId = null
                        if (_sessionExpired.value.not()) {
                            // Only set expired if we were previously signed in
                        }
                    }
                    else -> { /* Loading state */ }
                }
            }
        }
    }

    // --- Auth Methods ---

    suspend fun signInEmail(email: String, password: String) {
        client.auth.signInWith(Email) {
            this.email = email
            this.password = password
        }
    }

    suspend fun signUpEmail(email: String, password: String) {
        client.auth.signUpWith(Email) {
            this.email = email
            this.password = password
        }
    }

    suspend fun resetPassword(email: String) {
        client.auth.resetPasswordForEmail(email)
    }

    suspend fun resendConfirmation(email: String) {
        client.auth.resendEmail(io.github.jan.supabase.gotrue.providers.builtin.Email.Config.ResendType.SIGNUP, email)
    }

    suspend fun signInWithGoogle(idToken: String) {
        client.auth.signInWith(IDToken) {
            this.idToken = idToken
            this.provider = Google
        }
    }

    suspend fun signOut() {
        client.auth.signOut()
        cachedUserId = null
        _isSignedIn.value = false
        _currentUserEmail.value = null
        _sessionExpired.value = false
    }

    suspend fun restoreSession() {
        try {
            client.auth.retrieveUser()
            client.auth.refreshCurrentSession()
            _isSignedIn.value = client.auth.currentSessionOrNull() != null
            _currentUserEmail.value = client.auth.currentUserOrNull()?.email
        } catch (e: Exception) {
            Log.w(TAG, "Session restore failed: ${e.message}")
            _isSignedIn.value = false
        }
    }

    suspend fun deleteAccount() {
        // Call edge function to delete all data + auth user
        client.postgrest.rpc("delete_user_account")
        signOut()
    }

    // --- Push Token Registration ---

    @Serializable
    private data class UserDeviceRow(
        @SerialName("user_id") val userId: String,
        @SerialName("device_token") val deviceToken: String,
        @SerialName("device_name") val deviceName: String? = null,
        val platform: String = "android",
        @SerialName("last_seen_at") val lastSeenAt: String? = null
    )

    suspend fun registerDeviceToken(token: String, deviceName: String? = null) {
        if (!_isSignedIn.value) return
        val userId = cachedUserId ?: return
        try {
            val row = UserDeviceRow(
                userId = userId,
                deviceToken = token,
                deviceName = deviceName,
                platform = "android",
                lastSeenAt = java.time.Instant.now().toString()
            )
            client.postgrest.from("user_devices").upsert(row) {
                onConflict = "device_token"
            }
        } catch (e: Exception) {
            Log.w(TAG, "registerDeviceToken failed: ${e.message}")
        }
    }

    // --- Sync Methods ---

    fun scheduleDebouncedSync() {
        debouncedSyncJob?.cancel()
        debouncedSyncJob = scope.launch {
            delay(SYNC_DEBOUNCE_MS)
            performSync()
        }
    }

    fun syncOnForeground() {
        scope.launch { performSync() }
    }

    private suspend fun performSync() {
        if (!_isSignedIn.value || !_isConnected.value) return

        syncMutex.withLock {
            _isSyncing.value = true
            _syncError.value = null

            try {
                pushPendingItems()
                pullRemoteChanges()
                secureStorageService.setLastSyncTimestamp(System.currentTimeMillis())
            } catch (e: Exception) {
                handleSyncError(e)
            } finally {
                _isSyncing.value = false
            }
        }
    }

    private suspend fun pushPendingItems() {
        val pendingItems = trackerItemDao.getItemsBySyncStatus(SyncStatus.PENDING)
        if (pendingItems.isEmpty()) return

        withRetry("push") {
            val rows = pendingItems.map { item ->
                TrackerItemRow(
                    id = item.id,
                    name = item.name,
                    categoryId = item.categoryId,
                    reminderIntervalDays = item.reminderIntervalDays,
                    sortOrder = item.sortOrder,
                    iconName = item.iconName,
                    isArchived = item.isArchived
                )
            }

            client.postgrest.from("tracker_items").upsert(rows)

            // Mark items as synced
            pendingItems.forEach { item ->
                trackerItemDao.update(item.copy(syncStatus = SyncStatus.SYNCED))
            }
        }
    }

    private suspend fun pullRemoteChanges() {
        val lastSync = secureStorageService.getLastSyncTimestamp()

        withRetry("pull") {
            val remoteItems = if (lastSync > 0) {
                client.postgrest.from("tracker_items")
                    .select()
                    .decodeList<TrackerItemRow>()
            } else {
                client.postgrest.from("tracker_items")
                    .select()
                    .decodeList<TrackerItemRow>()
            }

            for (remoteRow in remoteItems) {
                val localItem = trackerItemDao.getItemById(remoteRow.id)
                if (localItem == null) {
                    // New item from server — insert
                    trackerItemDao.insert(
                        TrackerItem(
                            id = remoteRow.id,
                            name = remoteRow.name,
                            categoryId = remoteRow.categoryId,
                            reminderIntervalDays = remoteRow.reminderIntervalDays,
                            sortOrder = remoteRow.sortOrder,
                            iconName = remoteRow.iconName,
                            isArchived = remoteRow.isArchived,
                            syncStatus = SyncStatus.SYNCED
                        )
                    )
                } else if (localItem.syncStatus == SyncStatus.SYNCED) {
                    // No local changes — safe to overwrite with remote
                    trackerItemDao.update(
                        localItem.copy(
                            name = remoteRow.name,
                            categoryId = remoteRow.categoryId,
                            reminderIntervalDays = remoteRow.reminderIntervalDays,
                            sortOrder = remoteRow.sortOrder,
                            iconName = remoteRow.iconName,
                            isArchived = remoteRow.isArchived,
                            syncStatus = SyncStatus.SYNCED
                        )
                    )
                } else {
                    // Local has pending changes — conflict detection via updatedAt
                    // Keep local version, mark as pending for next push
                    Log.d(TAG, "Conflict detected for item ${remoteRow.id}, keeping local version")
                }
            }
        }
    }

    // --- Retry Logic ---

    private suspend fun <T> withRetry(operation: String, block: suspend () -> T): T {
        var lastException: Exception? = null

        for (attempt in 0 until MAX_RETRY_ATTEMPTS) {
            try {
                return block()
            } catch (e: Exception) {
                lastException = e
                val errorMessage = e.message ?: ""

                // Check for auth errors — don't retry, trigger re-auth
                if (errorMessage.contains("401") || errorMessage.contains("403") ||
                    errorMessage.contains("JWT", ignoreCase = true)) {
                    _sessionExpired.value = true
                    throw e
                }

                if (attempt < MAX_RETRY_ATTEMPTS - 1) {
                    Log.w(TAG, "Sync $operation attempt ${attempt + 1} failed, retrying in ${RETRY_DELAYS_MS[attempt]}ms: ${e.message}")
                    delay(RETRY_DELAYS_MS[attempt])
                }
            }
        }

        throw lastException ?: Exception("Sync $operation failed after $MAX_RETRY_ATTEMPTS attempts")
    }

    private fun handleSyncError(error: Exception) {
        val message = error.message ?: "Unknown sync error"
        Log.e(TAG, "Sync error: $message")
        _syncError.value = when {
            message.contains("401") || message.contains("403") -> {
                _sessionExpired.value = true
                "Session expired. Please sign in again."
            }
            message.contains("network", ignoreCase = true) ||
            message.contains("connection", ignoreCase = true) ->
                "Sync failed: No network connection. Changes saved locally."
            else -> "Sync failed: $message"
        }
    }
}
